core/src/Kokkos_GraphNode.hpp

```diff
@@ -246,6 +247,51 @@ class GraphNodeRef {
                       std::forward<Functor>(functor));
   }
 
+#if defined(KOKKOS_ENABLE_CUDA) ||                                           \
+    (defined(KOKKOS_ENABLE_HIP) && defined(KOKKOS_IMPL_HIP_NATIVE_GRAPH)) || \
+    (defined(KOKKOS_ENABLE_SYCL) && defined(SYCL_EXT_ONEAPI_GRAPH))
+  template <class Functor,
+            typename = std::enable_if_t<std::is_invocable_r_v<
+                void, const Kokkos::Impl::remove_cvref_t<Functor>,
+                const ExecutionSpace&>>>
+#if defined(KOKKOS_ENABLE_CUDA)
+  auto cuda_capture(const ExecutionSpace& exec, Functor&& functor) const {
+    if constexpr (std::is_same_v<ExecutionSpace, Kokkos::Cuda>) {
+#elif defined(KOKKOS_ENABLE_HIP) && defined(KOKKOS_IMPL_HIP_NATIVE_GRAPH)
+  auto hip_capture(const ExecutionSpace& exec, Functor&& functor) const {
+    if constexpr (std::is_same_v<ExecutionSpace, Kokkos::HIP>) {
+#elif defined(KOKKOS_ENABLE_SYCL) && defined(SYCL_EXT_ONEAPI_GRAPH)
+  auto sycl_capture(const ExecutionSpace& exec, Functor&& functor) const {
+    if constexpr (std::is_same_v<ExecutionSpace, Kokkos::SYCL>) {
+#endif
+      using capture_t = Kokkos::Impl::GraphNodeCaptureImpl<
+          ExecutionSpace, Kokkos::Impl::remove_cvref_t<Functor>>;
+      using return_t = GraphNodeRef<ExecutionSpace, capture_t, GraphNodeRef>;
+
+      auto graph_ptr = m_graph_impl.lock();
+      KOKKOS_EXPECTS(bool(graph_ptr))
+
+      auto rv = Kokkos::Impl::GraphAccess::make_graph_node_ref(
+          m_graph_impl,
+          Kokkos::Impl::GraphAccess::make_node_shared_ptr<
+              typename return_t::node_impl_t>(
+              m_node_impl->execution_space_instance(),
+              Kokkos::Impl::_graph_node_capture_ctor_tag{},
+              std::forward<Functor>(functor),
+              Kokkos::Impl::_graph_node_predecessor_ctor_tag{}, *this));
+
+      // Add the node itself to the backend's graph data structure, now that
+      // everything is set up.
+      graph_ptr->add_node(exec, rv.m_node_impl);
+      // Add the predecessor we stored in the constructor above in the
+      // backend's data structure, now that everything is set up.
+      graph_ptr->add_predecessor(rv.m_node_impl, *this);
+      KOKKOS_ENSURES(bool(rv.m_node_impl))
+      return rv;
+    }
```

Comment:
How about this: 
```suggestion
    } else {
       Kokkos::abort("Some appropriate message");
    }
```

Reply:
Or actually how about replacing the if constexrp above with `static_assert()`

Reply:
It tried to be consistent with what we did in #6904:
https://github.com/kokkos/kokkos/blob/ca8d31a245bc553b4de6d48e62ce8efe0c336154/core/src/Kokkos_Graph.hpp#L185-L201

And to quote @dalg24 at the time (https://github.com/kokkos/kokkos/pull/6904#discussion_r1767996063):
> They'll get a compile error if you don't return anything.